src/commands/login.rs

```diff
@@ -10,9 +13,57 @@ pub struct LoginSubcommand {
     pub token: Option<String>,
 }
 
+#[derive(Debug, Deserialize)]
+struct DeviceCodeResponse {
+    device_code: String,
+    user_code: String,
+    verification_uri: String,
+    expires_in: u64,
+    interval: u64,
+}
+
+#[derive(Deserialize)]
+struct DeviceCodeAuth {
+    access_token: String,
+    token_type: String,
+    scope: String,
+}
+
+#[derive(Deserialize)]
+#[serde(untagged)]
+enum AuthResponse {
+    Ok(DeviceCodeAuth),
+    Err { error: String },
+}
+
+fn await_github_auth(device_code_response: DeviceCodeResponse) -> anyhow::Result<DeviceCodeAuth> {
+    let client = reqwest::blocking::Client::new();
+
+    thread::sleep(time::Duration::from_secs(device_code_response.interval));
+
+    let response = client
+        .post("https://github.com/login/oauth/access_token")
+        .header("accept", "application/json")
+        .json(&serde_json::json!({
+            "client_id": "7bd503594a0f9a9f7ed3",
+            "device_code": device_code_response.device_code,
+            "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
+        }))
+        .send()?
+        .json::<AuthResponse>()?;
+
+    match response {
+        AuthResponse::Ok(auth) => Ok(auth),
+        AuthResponse::Err { error } => match error.as_ref() {
+            "authorization_pending" => await_github_auth(device_code_response),
+            _ => Err(format_err!("Oauth request error: {}", error)),
+        },
+    }
+}
+
 impl LoginSubcommand {
     pub fn run(self) -> anyhow::Result<()> {
-        let token = match self.token {
+        /*let token = match self.token {
```

Comment:
Talked about this in Discord, but we'll wanna preserve the existing auth codepath.

Reply:
Auth is now done on a per registry basis. If the target index doesn't have an `oauth_id` then wally will default to asking for an api key. If an `oauth_id` does exist then it will use that and present an oauth flow.